File: python/silapay_connector/silapay_connector/providers/silapay.py
import json
import os

import requests

BASE_URL = "https://api.silapay.pro/v1"


class SilapayProvider:
    def __init__(self, api_key=None, secret_key=None, base_url=BASE_URL, timeout=30):
        api_key = api_key or os.getenv("SILAPAY_API_KEY")
        secret_key = secret_key or os.getenv("SILAPAY_SECRET_KEY")

        if not api_key or not secret_key:
            raise ValueError(
                "Silapay: defina SILAPAY_API_KEY e SILAPAY_SECRET_KEY no ambiente."
            )

        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update(
            {
                "accept": "application/json",
                "content-type": "application/json",
                "api-key": api_key,
                "secret-key": secret_key,
            }
        )

    def pix(self, dados):
        body = {"paymentMethod": "pix", **dados}
        return self._request("POST", "/transactions", json_body=body)

    def boleto(self, dados):
        body = {"paymentMethod": "billet", **dados}
        return self._request("POST", "/transactions", json_body=body)

    def cartao(self, dados):
        body = {"paymentMethod": "creditCard", **dados}
        return self._request("POST", "/transactions", json_body=body)

    def saldo(self):
        return self._request("GET", "/user/finance/balance")

    def _request(self, method, path, json_body=None):
        try:
            response = self.session.request(
                method=method,
                url=f"{self.base_url}{path}",
                json=json_body,
                timeout=self.timeout,
            )
            response.raise_for_status()

            if not response.content:
                return None

            return response.json()
        except requests.HTTPError as err:
            raise create_api_error("Silapay", err) from err


def create_api_error(provider, err):
    response_data = parse_response_data(err.response)
    message = f"{provider} API error: {serialize_response_data(response_data)}"
    api_error = RuntimeError(message)
    api_error.response = err.response
    api_error.response_data = response_data
    return api_error


def parse_response_data(response):
    try:
        return response.json()
    except ValueError:
        return response.text


def serialize_response_data(data):
    if isinstance(data, str):
        return data

    return json.dumps(data, ensure_ascii=False)

File: typescript/silapay-connector/src/providers/silapay.ts
import axios, { AxiosError, AxiosInstance } from "axios";

import type { ApiError, ConnectorOptions, PaymentProvider } from "../types";

const BASE_URL = "https://api.silapay.pro/v1";

export class SilapayProvider implements PaymentProvider {
    private readonly client: AxiosInstance;

    constructor(options: ConnectorOptions = {}) {
        const apiKey = options.apiKey || process.env.SILAPAY_API_KEY;
        const secretKey = options.secretKey || process.env.SILAPAY_SECRET_KEY;

        if (!apiKey || !secretKey) {
            throw new Error("Silapay: defina SILAPAY_API_KEY e SILAPAY_SECRET_KEY no ambiente.");
        }

        this.client = axios.create({
            baseURL: options.baseURL || BASE_URL,
            timeout: options.timeout || 30000,
            headers: {
                accept: "application/json",
                "content-type": "application/json",
                "api-key": apiKey,
                "secret-key": secretKey,
            },
        });
    }

    async pix(dados: Record<string, unknown>): Promise<unknown> {
        return this.request("post", "/transactions", {
            paymentMethod: "pix",
            ...dados,
        });
    }

    async boleto(dados: Record<string, unknown>): Promise<unknown> {
        return this.request("post", "/transactions", {
            paymentMethod: "billet",
            ...dados,
        });
    }

    async cartao(dados: Record<string, unknown>): Promise<unknown> {
        return this.request("post", "/transactions", {
            paymentMethod: "creditCard",
            ...dados,
        });
    }

    async saldo(): Promise<unknown> {
        return this.request("get", "/user/finance/balance");
    }

    private async request(
        method: "get" | "post",
        url: string,
        data?: Record<string, unknown>
    ): Promise<unknown> {
        try {
            const response = await this.client.request({ method, url, data });
            return response.data;
        } catch (error) {
            throw createApiError("Silapay", error);
        }
    }
}

function createApiError(provider: string, error: unknown): Error {
    if (!isAxiosError(error) || typeof error.response === "undefined") {
        return error instanceof Error ? error : new Error(String(error));
    }

    const responseData = error.response?.data;
    const apiError = new Error(
        `${provider} API error: ${serializeResponseData(responseData)}`
    ) as ApiError;

    apiError.response = error.response;
    apiError.responseData = responseData;

    return apiError;
}

function serializeResponseData(data: unknown): string {
    if (typeof data === "string") {
        return data;
    }

    return JSON.stringify(data);
}

function isAxiosError(error: unknown): error is AxiosError {
    return axios.isAxiosError(error);
}
